lib: cobs: implement streaming cobs variant

mkschreder opened the pull request:
## Summary
This PR adds a streaming (incremental) COBS encode/decode API to Zephyr’s COBS utility, along with comprehensive unit tests.

## Why
Many transports deliver data in small, arbitrary chunks (for example, UART RX buffers). In these cases, treating COBS encoding/decoding as a single “whole buffer in / whole buffer out” operation is awkward. This paves the way for implementing new cobs based uart slip network interface which I will push in a separate PR. 

## What’s added
- **New state-machine based streaming APIs** for COBS:
  - Initialize/reset encoder/decoder state
  - Encode to an output buffer across multiple calls
  - Decode incrementally until a frame delimiter is encountered
- **Tests covering streaming behavior**, including:
  - Chunked/fragmented input handling
  - Boundary conditions (notably delimiter-at-boundary cases)
  - Roundtrip encode/decode validation

## Compatibility
- The existing COBS APIs remain intact; this change adds new capabilities without removing or redefining existing behavior.

## Testing
- Extended `tests/lib/cobs` with streaming-focused test coverage.

mkschreder committed:
lib: cobs: implement streaming cobs variant

Many transports deliver data in small, arbitrary chunks (for example,
UART RX buffers). In these cases, treating COBS encoding/decoding as a
single “whole buffer in / whole buffer out” operation is awkward.

This paves the way for being able to add highly efficient cobs-based
interfaces that operate on chunks instead of whole buffers. We avoid
creating a new buffer just to hold incomplete frame and instead maintain
state as we are receiving more data.

Encoding is a bit more complex since cobs requires lookahead when
encoding, but this is not a problem because we have access to full data
when we encode so we can do lookahead when needed.

Signed-off-by: Martin Schröder <info@swedishembedded.com>

mkschreder commented on include/zephyr/data/cobs.h:
```diff
@@ -86,8 +86,109 @@ int cobs_encode(struct net_buf *src, struct net_buf *dst, uint32_t flags);
+ * Non-destructive: tracks read position without modifying source.
+ */
+struct cobs_encode_state {
+	struct net_buf *src_frag;   /* Current source fragment being read */
```
Fixed.

pdgendt commented on lib/utils/cobs.c:
```diff
@@ -1,104 +1,638 @@
+	self->src_offset = 0;
+	self->block_code = 0;
+	self->block_pos = 0;
+}
```
`init` and `reset` are identical, keep one. The argument shouldn't be NULL (as the doxygen suggests, so change that to an assert.

```suggestion
	__ASSERT_NO_MSG(self != NULL);

	self->src_frag = NULL;
	self->src_offset = 0;
	self->block_code = 0;
	self->block_pos = 0;
}
```

pdgendt commented on lib/utils/cobs.c:
```diff
@@ -1,104 +1,638 @@
+/* Skip empty fragments to find next available data */
+static inline void skip_empty_fragments(struct cobs_encode_state *self)
+{
+	while (self->src_frag && self->src_offset >= self->src_frag->len) {
```
No implicit boolean conversions (same for others)

```suggestion
	while (self->src_frag != NULL && self->src_offset >= self->src_frag->len) {
```

pdgendt reviewed (changes requested):
We shouldn't have (or need to maintain) different implementations for the streaming mechanism vs block based.

Ideally the `cobs_decode/cobs_encode` functions simply use the streaming with a stack allocated state.

The streaming decoder/encoders also don't have support for custom delimiters...

mkschreder committed:
lib: cobs: refactor to support custom delimiter and consolidate

- Add back support for custom delimiter everywhere
- Consolidate cobs_encode/cobs_decode to call streaming api
- Add additional test cases for 0x00 block and fix that as well
- Remove implicit conversions to boolean

Signed-off-by: Martin Schröder <info@swedishembedded.com>

mkschreder commented on lib/utils/cobs.c:
```diff
@@ -1,104 +1,638 @@
+					      const uint32_t flags,
+					      const uint8_t delimiter)
+{
+	if (!(flags & COBS_FLAG_TRAILING_DELIMITER)) {
```
Fixed as well

hlord2000 commented on lib/utils/cobs.c:
```diff
@@ -1,104 +1,625 @@
+/**
+ * @brief Validate delimiter value
+ *
+ * COBS overhead bytes range from 0x01 to 0xFF. Only 0x00 is guaranteed
```
there is no requirement for the delimiter to be zero, this is just what is commonly done. it can be any value.
